presto-spi/src/main/java/com/facebook/presto/spi/block/BlockBuilder.java

```diff
@@ -62,6 +62,15 @@ default BlockBuilder writeObject(Object value)
     }
 
     /**
+     * Return a writer to the current entry. The caller can operate on the returned caller to incrementally build the object. This is generally more efficient than
+     * building the object elsewhere and call writeObject afterwards because a large chunk of memory could potentially be unnecessarily copied in this process.
+     */
+    default BlockBuilder nextObjectWriter()
```

Comment:
Putting this function in a specific BlockBuilder and downcast when needed is not an option. All three implementations added in this pull request needs to implement this class although the only one that "really" need it is ArrayBlock.